_blogposts/2025-04-05-introducing-unified-operators.mdx

```diff
@@ -0,0 +1,268 @@
+// This has type: float
+//   But it's being compared to something of type: int
+// 
+//   You can only compare things of the same type.
+//   
+//   You can convert float to int with Belt.Float.toInt.
+//   If this is a literal, try a number without a trailing dot (e.g. 20).
+```
+
+Because ReScript only intentionally supports monomorphic operations, `(int, int) => int` in this case. Users have to perform type conversions explicitly where necessary.
+
+While it's tempting to allow full operator overloading or polymorphism like JavaScript or TypeScript, we intentionally avoid it to preserve predictable type inference and runtime performance guarantees.
+
+However, comparisons are actually exceptional ones. Let's summon the polymorphism.
+
+```res
+let compare_poly = (a, b) => a < b
+```
+
+```js
+import * as Primitive_object from "./stdlib/Primitive_object.js";
+
+let compare_poly = Primitive_object.lessthan;
+```
+
+As both operands `a` and `b` are inferred as "open type", it turned it into a "runtime primitive" that takes any type and performs a struct comparison at runtime.
+
+This is a design decision to support comparisons for arbitrary record or tuple types, but it is not ideal. The runtime primitive is not well optimized and too expensive for common arithmetic operations.
+
+## Unified operators
+
+Unlike polymorphic operators, unified operators don't use runtime primitives at all. Instead, they modify the compiler to translate specific operators to existing compile-time primitives.
+
+More specifically, the following rules are added to the primitive translation process.
+
+> Before handling a primitive, if the primitive operation matches the form of `lhs -> rhs -> result` or `lhs -> result`
+>
+> 1. If the `lhs` type is a primitive type, unify the `rhs` and the `result` type to the `lhs` type.
+>   2. If the `lhs` type is not a primitive type but the `rhs` type is, unify `lhs` and the `result` type to the `rhs` type.
+>   3. If both `lhs` type and `rhs` type is not a primitive type, unify the whole types to the `int`.
+
+It changes the type inference like
+
+```res
+let t1 = 1 + 2     // => (int, int) => int
+let t2 = 1. + 2.   // => (float, float) => float
+let t3 = "1" + "2" // => (string, string) => string
+let t4 = 1n + 2n   // => (bigint, bigint) => bigint
+
+let fn1 = (a, b) => a + b        // (int, int) => int
+let fn2 = (a: float, b) => a + b // (float, float) => float 
+let fn3 = (a, b: float) => a + b // (float, float) => float
+
+let inv1 = (a: int, b: float) => a + b  // => (int, int) => int
+//                                   ^ error: cannot apply float here, expected int
+```
+
+Then, in IR, it is translated to the corresponding compile-time primitive based on the unified type.
+
+This approach has been inspired by an awesome language [F#](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/symbol-and-operator-reference/arithmetic-operators#operators-and-type-inference), which is also originated from OCaml.
```

Comment:
This approach is inspired by the awesome language [F#](https://learn.microsoft.com/en-us/dotnet/fsharp/language-reference/symbol-and-operator-reference/arithmetic-operators#operators-and-type-inference), which also originates from OCaml.